Added Argument parser for DPSDGYM

Tecnarca opened the pull request:
Modified the command line interface to take as input only specific datasets, epsilons and metrics to run.
Moved the default array for metrics and epsilons in config.py
Updated README.md to match current input system.

Possible extension: allow one to select models to be run (from the ones in config.py) and let one select the mlflow run name (currently defaults to "test").

eedeleon commented on tests/dpsdgym/README.md:
````diff
@@ -56,17 +56,17 @@ Make sure you have the following installed in a conda env, before running the ev
+Flags can be specified to specify which metrics, epsilon and datasets should be used to run the evaluation:
 ```
-python main.py wasserstein pmse ml_eval sra
+python main.py -d bank adult -e 0.01 0.1 1 -m pmse ml_eval wasserstein
````
Part of the reason we avoided this approach is that mlflow did not properly parse the arguments for the remote scenario.

Can you run these changes with mlflow run "tests/dpsdgym" -P {param}={val} to verify the changes are compatible?

eedeleon replied:
I am going to assume this comment is not a blocker since those configuration points were not there to begin with

eedeleon commented on tests/dpsdgym/main.py:
```diff
@@ -3,6 +3,9 @@
 import json
+import argparse
+import textwrap
+
```
We will need to update the MLproject file in this directory to reflect the changes to main.py's CLI interface

Tecnarca commented:
I have corrected the argument parsing for MLFlow by parsing further the arguments when multiple values come in a single string as MLFlow was doing.
The command: 
`mlflow run "tests/dpsdgym" -P dataset="bank adult mushroom" -P epsilons="1 2" -P eval_metrics="mse ml_eval"`
Now runs the project correctly. I've updated the MLProject file accordingly to reflect this change.

I forgot to mention why this change is needed: it allows to specify epsilons prerun and this is is going to be important when we will introduce a checkpointing system in the future (otherwise we would not be able to run automatically different experiments, we would have to modify conf.py manually). It also allows to run multiple metrics and datasets in the same run, when needed.

eedeleon commented:
/AzurePipeline run

eedeleon commented on tests/dpsdgym/benchmark.py:
```diff
@@ -0,0 +1,41 @@
+	    if 'ml_eval' in metric:
+	        results = run_ml_eval(data_dicts, epsilons, run_name)
+	        
+	        if 'sra' in metric:
```
Is this supposed to be tabbed in?

eedeleon commented on tests/dpsdgym/main.py:
```diff
@@ -1,52 +1,34 @@
-from evaluate import run_ml_eval, run_wasserstein, run_pMSE
+    args = parser.parse_args()
+    if isinstance(args.dataset[0], str) and len(args.dataset)==1:
+        args.dataset = args.dataset[0].split()
```
This is ok for now but comma separated strings will probably be better. Dataset names are more likely to have a space than a comma

eedeleon commented on tests/dpsdgym/main.py:
```diff
@@ -1,52 +1,34 @@
-from warnings import simplefilter
-simplefilter(action='ignore', category=FutureWarning)
+    if isinstance(args.epsilon[0], str) and len(args.epsilon)==1:
+        args.epsilon = args.epsilon[0].split()
```
the space comment is less of an issue here :)

eedeleon commented on tests/dpsdgym/main.py:
```diff
@@ -1,52 +1,34 @@
-        json.dump(results, f)
-    mlflow.log_artifact("artifact.json", "results.json")
+    if isinstance(args.metric[0], str) and len(args.metric)==1:
+        args.metric = args.metric[0].split()
```
This can probably be a util function args.metric = split_if_needed(args.metric)

eedeleon commented on tests/dpsdgym/synthesis.py:
```diff
@@ -39,13 +39,13 @@ def run_synthesis(synthesis_args):
     return (n, d, str(e), sampled)
 
-def run_all_synthesizers(datasets, epsilons):
+def run_all_synthesizers(datasets, epsilons, save_models_path, run_name):
```
Can the new parameters be optional with defaults?

eedeleon reviewed (approved):
These are great additions. Thank you so much for taking the time to add so many configuration and extension point possibilities to the Eval.

I think the only part that might leave the eval outputs hard to track is sra use + ml_eval results being modified. This change is ok for now though hence the approve since the comments are mostly around style
